Bluetooth: Mesh: Add missing comp pages to LCD mod

Andrewpini opened the pull request:
- Adds support for all composition data pages to the
Large Composition Data model.

- Adds test for verifying split get operation for all
composition data pages.

Fixes #63629

PavelVPV commented on subsys/bluetooth/mesh/large_comp_data_srv.c:
```diff
@@ -59,24 +59,54 @@ static int handle_large_comp_data_get(struct bt_mesh_model *model, struct bt_mes
 	bt_mesh_model_msg_init(&rsp, OP_LARGE_COMP_DATA_STATUS);
-
-	if (page != 0U) {
+	if (page >= 130U && IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_2) &&
```
Can all this code be moved into `access.c` and reused in `cfg_srv.c` with `offset` == 0?

Andrewpini replied:
I will investigate the possibillity. I think I will push all other resolved comments first, then push a fix for this later so that we can keep the momentum up on the PR.

PavelVPV replied:
by later you mean in a separate PR?

Andrewpini replied:
No, I've added a suggestion commit now

PavelVPV replied:
No, sorry, I don't like the new function. It adds a lot of models specific dependencies to the access layer. What I was worried about is the bunch of if-else cases for page selection and then reading the page depending on the page number and dirty flag. But this change now create the opposite situation. Lets drop the last commit and do the refactoring some time later.

Andrewpini replied:
I changed it to only include the page parsing. At least a minor improvment IMO.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_lcd.c:
```diff
@@ -255,32 +290,40 @@ static void test_cli_split_comp_data_request(void)
-	rsp_equals_local_data_assert(SRV_ADDR, &srv_rsp_1, &local_comp, page, offset, total_size,
-				     &prev_len);
+	ASSERT_OK(bt_mesh_large_comp_data_get(0, SRV_ADDR, comp_page, offset, &srv_rsp_1));
+
```
Please remove the added empty line here and below. Lets move style changes to a separate commit/PR so it is easier to review.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1375,6 +1375,14 @@ config BT_MESH_PRIV_BEACON_CLI
+	help
+	  Buffer used to temporarily hold composition data pages
+	  during flash operations. Should reflect the size of the
+	  largest composition data page present in the application.
```
Please add that this buffer is allocated on stack. Also add that after DFU this buffer should be big enough to restore previously stored page, otherwise the page won't be fully restored.

alxelax replied:
Hm. If this is allocated on stack, should stack size be increased either relatively?

Andrewpini replied:
@alxelax Hmm... not sure.

alxelax replied:
Did you try to pass provisioning procedure with enabled settings in this new changes on real device?

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -623,8 +602,15 @@ static int bt_mesh_comp_data_get_page_1(struct net_buf_simple *buf)
 	for (i = 0; i < comp->elem_count; i++) {
-		if (net_buf_simple_tailroom(buf) <
-		    (page1_elem_size(&comp->elem[i]) + BT_MESH_MIC_SHORT)) {
+
```
empty line

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -639,70 +625,74 @@ static int bt_mesh_comp_data_get_page_1(struct net_buf_simple *buf)
 				 */
-				LOG_DBG("Record 0x%04x didn't fit in the Data field",
-					i);
+				LOG_DBG("Record 0x%04x didn't fit in the Data field", i);
```
This function and its siblings are used to store cdp on flash. If LCD is present, but  `CONFIG_BT_MESH_COMP_PST_BUF_SIZE` isn't big enough, it will silently stop here and only part of the cdp will be stored. I think it would make sense to add a check after all CDPs are initialized or in `bt_mesh_comp_store` that would warn if `CONFIG_BT_MESH_COMP_PST_BUF_SIZE` isn't big enough to hold the largest CDP.

PavelVPV commented on tests/bsim/bluetooth/mesh/tests_scripts/large_comp_data/get_comp2_data_split .sh:
```diff
@@ -19,11 +19,25 @@ source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
 conf=prj_mesh1d1_conf
-RunTest mesh_lcd_test_split_comp_data \
-	lcd_cli_split_comp_data_request lcd_srv_comp_data_status_respond
+overlay=overlay_pst_conf
```
I'm uncertain if we need to change these tests to run with settings subsystem. Was there a particular need for this? We can always instantiate a separate `bst_test_instance` instance in the same test file wrapped into `if defined(CONFIG_BT_SETTINGS)` and added to `overlay_pst_conf`.

Andrewpini replied:
Basically I need a functioning flash so that I can verify that the parsing functionality for fetching comp data pages from persistent storage works as expected. To do so I need the settings subsystem right?

PavelVPV replied:
Ok, sorry, I missed that you changed old tests. Than it is fine!

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1375,6 +1375,14 @@ config BT_MESH_PRIV_BEACON_CLI
+	int "Composition Data Page persistence buffer size"
+	default 100
+	help
+	  Buffer used to temporarily hold composition data pages
+	  during flash operations. Should reflect the size of the
+	  largest composition data page present in the application.
```
```suggestion
	  Buffer used to temporarily hold Composition Data Pages
	  during flash operations. Should reflect the size of the
	  largest Composition Data Page present in the application.
```

Andrewpini commented:
> @Andrewpini, why do you use persistent configuration in bsim tests if every new test is run with empty flash? Doesn't seem to me, it has any sense.

I need 

> @Andrewpini, why do you use persistent configuration in bsim tests if every new test is run with empty flash? Doesn't seem to me, it has any sense.

Discussed and resolved in person

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1375,6 +1375,17 @@ config BT_MESH_PRIV_BEACON_CLI
+	help
+	  Stack allocated buffer used to temporarily hold Composition
+	  Data Pages during flash operations. Should reflect the size
+	  of the largest Composition Data page present in the application.
```
```suggestion
	  of the largest Composition Data Page present in the application.
```

mia-ko commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1375,6 +1375,17 @@ config BT_MESH_PRIV_BEACON_CLI
+	  Stack allocated buffer used to temporarily hold Composition
+	  Data Pages during flash operations. Should reflect the size
+	  of the largest Composition Data page present in the application.
+	  It is important to note that after device firmware upgrade this
+	  buffer should still be large enough to restore previously stored
+	  pages.
```
```suggestion
	  Note that this buffer should still be large enough to restore previously stored
	  pages after a performed device firmware update.
```

akredalen reviewed (commented):
Looks good to me 👍 Will approve when all remaining comments have been resolved.

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2212,20 +2203,87 @@ int bt_mesh_comp_data_get_page(struct net_buf_simple *buf, size_t page, size_t o
 	for (int i = 0; i < ARRAY_SIZE(comp_data_pages); i++) {
+		if (bt_mesh_comp_page_size(i) > CONFIG_BT_MESH_COMP_PST_BUF_SIZE) {
+			LOG_WRN("CDP%d is larger than the CDP persistence buffer. "
+				"Please increase the CDP persistence buffer size.",
```
Print required size?

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2212,20 +2203,87 @@ int bt_mesh_comp_data_get_page(struct net_buf_simple *buf, size_t page, size_t o
+{
+	const struct bt_mesh_comp *comp;
+	const struct bt_mesh_elem *elem;
+	size_t size = 10;
```
Could you add comment what 10 means? Seems like magic digit.

alxelax commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2212,20 +2203,87 @@ int bt_mesh_comp_data_get_page(struct net_buf_simple *buf, size_t page, size_t o
+
+	comp = bt_mesh_comp_get();
+
+	for (i = 0; i < comp->elem_count; i++) {
```
```suggestion
	for (int i = 0; i < comp->elem_count; i++) {
```

alxelax commented on tests/bsim/bluetooth/mesh/src/test_lcd.c:
```diff
@@ -50,6 +51,41 @@ const struct bt_mesh_model_op dummy_op[] = {
 };
 
+static const uint8_t elem_offset2[3] = {4, 5, 6};
+static const uint8_t additional_data[2] = {100, 200};
```
Could you add comment what `additional_data` means or reference on spec?

alxelax reviewed (commented):
LGTM, just minor comments

Andrewpini committed:
Bluetooth: Mesh: Add missing comp pages to LCD mod

Adds support for all composition data pages to the
Large Composition Data model.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
tests: Bluetooth: Mesh: LCD test all comp pages

Adds test for verifying split get operation for all
composition data pages.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Common comp page parse func

Create common composition page parser function.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

jhedberg commented:
@PavelVPV please start making sure that any Mesh PRs for 3.5 have a linked bug report (GitHub issue). When I read the PR description it doesn't sound at all like a bug fix, but still you have added the "bug" label to this one.

PavelVPV commented:
Fixes #63629
